### ISSUE TITLE
bug: NullPointerException in MethodTypingContext

### ISSUE THREAD

**Issue by dya-tel** [INRIA/spoon#3740]
The first here line assumes that `typeParam` is never null:
https://github.com/INRIA/spoon/blob/3435facc2b7089119a0ac278a2e008fde7d7c6b7/src/main/java/spoon/support/visitor/MethodTypingContext.java#L173-L177
But the argument (in the last return statement) actually **can** be null because the declaration can be null is some cases:
https://github.com/INRIA/spoon/blob/3435facc2b7089119a0ac278a2e008fde7d7c6b7/src/main/java/spoon/support/visitor/AbstractTypingContext.java#L66-L71

This bug is somewhat related to #3733 and #3734 - it surfaced when I tried fixing the latter by not copying the parent when it doesn't exist. Now adaptTypeParameterReference gets an invalid (?) generic from a corrupted Object type reference because of #3733. This generic doesn't have a declaration and causes the above problem.

**Comment by monperrus** [comment]
Would you be able to propose a failing test case in a PR? Thanks!

**Comment by dya-tel** [comment]
Nope, sorry. No ideas on how to reproduce the NPE "properly". I could probably try making a synthetic test where the reference doesn't have a declaration, but this doesn't seem like a good test to me.

Actually, I'm not entirely sure that this NPE is possible in normal conditions. But it seems suspicious to me that a declaration (which usually can be null for a lot of the tree elements) is used without a null-check.
